CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h

```diff
@@ -0,0 +1,150 @@
+  TrackingRecHit2DHeterogeneous(TrackingRecHit2DHeterogeneous&&) = default;
+  TrackingRecHit2DHeterogeneous& operator=(TrackingRecHit2DHeterogeneous&&) = default;
+
+  TrackingRecHit2DSOAView* view() { return m_view.get(); }
+  TrackingRecHit2DSOAView const* view() const { return m_view.get(); }
+
+  auto nHits() const { return m_nHits; }
+
+  auto hitsModuleStart() const { return m_hitsModuleStart; }
+  auto hitsLayerStart() { return m_hitsLayerStart; }
+  auto phiBinner() { return m_hist; }
+  auto iphi() { return m_iphi; }
+
+  // only the local coord and detector index
+  cudautils::host::unique_ptr<float[]> localCoordToHostAsync(cudaStream_t stream) const;
+  cudautils::host::unique_ptr<uint16_t[]> detIndexToHostAsync(cudaStream_t stream) const;
+  cudautils::host::unique_ptr<uint32_t[]> hitsModuleStartToHostAsync(cudaStream_t stream) const;
+
+private:
+  static constexpr uint32_t n16 = 4;
+  static constexpr uint32_t n32 = 9;
+  static_assert(sizeof(uint32_t) == sizeof(float));  // just stating the obvious
+
+  unique_ptr<uint16_t[]> m_store16;  //!
+  unique_ptr<float[]> m_store32;     //!
+
+  unique_ptr<TrackingRecHit2DSOAView::Hist> m_HistStore;                        //!
+  unique_ptr<TrackingRecHit2DSOAView::AverageGeometry> m_AverageGeometryStore;  //!
+
+  unique_ptr<TrackingRecHit2DSOAView> m_view;  //!
+
+  uint32_t m_nHits;
+
+  uint32_t const* m_hitsModuleStart;  // needed for legacy, this is on GPU!
+
+  // needed as kernel params...
+  Hist* m_hist;
+  uint32_t* m_hitsLayerStart;
+  int16_t* m_iphi;
+};
+
+#include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+template <typename Traits>
+TrackingRecHit2DHeterogeneous<Traits>::TrackingRecHit2DHeterogeneous(uint32_t nHits,
+                                                                     pixelCPEforGPU::ParamsOnGPU const* cpeParams,
+                                                                     uint32_t const* hitsModuleStart,
+                                                                     cudaStream_t stream)
+    : m_nHits(nHits), m_hitsModuleStart(hitsModuleStart) {
+  auto view = Traits::template make_host_unique<TrackingRecHit2DSOAView>(stream);
+
+  view->m_nHits = nHits;
+  m_view = Traits::template make_device_unique<TrackingRecHit2DSOAView>(stream);
+  m_AverageGeometryStore = Traits::template make_device_unique<TrackingRecHit2DSOAView::AverageGeometry>(stream);
+  view->m_averageGeometry = m_AverageGeometryStore.get();
+  view->m_cpeParams = cpeParams;
+  view->m_hitsModuleStart = hitsModuleStart;
+
+  // if empy do not bother
+  if (0 == nHits) {
+    if
+#ifndef __CUDACC__
+        constexpr
+#endif
+        (std::is_same<Traits, cudaCompat::GPUTraits>::value) {
+      cudautils::copyAsync(m_view, view, stream);
+    } else {
+      m_view.reset(view.release());  // NOLINT: std::move() breaks CUDA version
+    }
```

Comment:
We could also use older "dialect" of metaprogramming to never expose the "else" branch for the CUDA version.

This makes me realize that the `reset()`+`release()` with the host-pinned and device `unique_ptr's` is dangerous because of the custom deleters (e.g. this kind of code would make the host memory to be released with `cudaFree()` leading to an exception to be thrown there), and could be prevented by not providing such functions.